### PR TITLE
Add `useParams` and `useRouterQuery` hooks

### PR DESCRIPTION
**Pull request by anteprimorac**
Related: #530 

### What are the changes and their implications?
This PR implements two hooks `useRouterParams` and `useRouterQuery`.

### Checklist

- [x] Tests added for changes
- [ ] User facing changes documented

### Breaking change: yes/no
No.

### Other information
Example of usage:

```ts
// app/locations/pages/locations/[postalCode].tsx
import { useRouterParams, useRouterQuery, useQuery } from 'blitz';

export const Locations = () => {
  const { postalCode } = useRouterParams();
  const { order, orderby } = useRouterQuery();
  const [locations] = useQuery(getLocations, { where: { postalCode }, orderBy: { [orderby]: order } });
  return locations;
};
```

### CODE REVIEW

File: packages/core/src/use-router-query.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import {useRouter} from 'next/router'
+import {parse} from 'url'
```
Comment (flybayer) [inline packages/core/src/use-router-query.ts:2]: Need to add `url` as a dependency of the `blitzjs/core` package

Comment (anteprimorac) [reply packages/core/src/use-router-query.ts:2]: `url` dependency is added. I've also rewritten the implementation of `useRouterParams` as you described, replaced `router.query` in the store example with new hooks and added integration test that tests if `/admin/products?order=asc` changes the order of products in DOM

> Lastly, before merging this we should resolve our discussion in #530 about number parsing. Maybe we can have `useParamsStr`, `useParamsInt`, and `useRouterQuery`

Let's discuss it in the issue.

**Comment by flybayer** [comment]
@anteprimorac can you also add at least one unit test for `useRouterParams`?

**Comment by Zeko369** [comment]
@anteprimorac Also rerun prettier with our config, it seems like some files weren't formatted correctly

**Comment by anteprimorac** [comment]
> @anteprimorac can you also add at least one unit test for `useRouterParams`?

I've changed store example to use `useRouterParams` here https://github.com/blitz-js/blitz/pull/574/commits/ba730b17fd648418d26ae082064c2a5bd1b07aad#diff-1b842ad42068e396b7c21d9376384d87R8 and integration test is run over it. Is that enough?

It's hard to write unit tests because it depends on `useRouter` hook from next.js. Not sure if they provide a mock for it.

> @anteprimorac Also rerun prettier with our config, it seems like some files weren't formatted correctly

🤔 it looks like there is a separate prettier configuration for each example. Going to fix it now. Thanks for pointing this.

File: packages/core/src/use-router-params.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+export function useRouterParams() {
+  const router = useRouter()
+  const query = useRouterQuery()
+
+  return Object.fromEntries(
+    Object.entries(router.query).filter(
+      ([key, value]) => typeof query[key] === 'undefined' || query[key] !== value,
+    ),
+  )
```
Comment (flybayer) [inline packages/core/src/use-router-params.ts:12]: This is the part that needs unit tests — you can extract it into a function outside useRouterParams, export it, and then test this.

Comment (anteprimorac) [reply packages/core/src/use-router-params.ts:12]: Yeah, for sure, that could be exported and tested. Thanks for pointing this out. I will make a unit test for that.

Comment (anteprimorac) [reply packages/core/src/use-router-params.ts]: @flybayer the test is added. Let's see does it pass 😄.

File: packages/core/package.json [hunk 3]
```diff
@@ -46,7 +46,8 @@
   "dependencies": {
     "pretty-ms": "6.0.1",
     "react-query": "1.3.3",
-    "serialize-error": "6.0.0"
+    "serialize-error": "6.0.0",
+    "url": "^0.11.0"
```
Comment (flybayer) [inline packages/core/package.json:50]: Change version to be pinned `"0.11.0"`

File: packages/core/src/use-router-params.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      }
+    }
+
+    return true
+  }
+
+  return value1 === value2
+}
+
+export function exportRouterParams(routerQuery: ParsedUrlQuery, query: ParsedUrlQuery) {
```
Comment (flybayer) [inline packages/core/src/use-router-params.ts:33]: This is minor, but let's rename to "extract" which is a good bit more clear about what's happening :)

```suggestion
export function extractRouterParams(routerQuery: ParsedUrlQuery, query: ParsedUrlQuery) {
```

File: packages/core/package.json (hunk 3 above)
Comment (flybayer) [inline packages/core/package.json:50]: Change version to be pinned "0.11.0"

File: packages/core/src/use-router-params.ts (hunk 4 above)
Comment (flybayer) [inline packages/core/src/use-router-params.ts:33]: This is minor, but let's rename to "extract" which is a good bit more clear about what's happening :)

```suggestion
export function extractRouterParams(routerQuery: ParsedUrlQuery, query: ParsedUrlQuery) {
```

File: packages/core/test/use-router-params.test.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        cat: 'category',
+        slug: ['example', 'multiple', 'slugs'],
+        empty: '',
+        queryArray: ['1', '123', ''],
+      }
+
+      const query = {
+        cat: 'somethingelse',
+        slug: ['query-slug'],
+        queryArray: ['1', '123', ''],
```
Comment (flybayer) [inline packages/core/test/use-router-params.test.ts:17]: There's one more case which isn't covered here which is the result `params` should not include a key that is in `query`, but not in `routerQuery`. So I think just add a key here to `query` and tests should still pass.

Makes sense? :)

File: README.md [hunk 6]
```diff
@@ -270,6 +270,7 @@ Thanks to these wonderful people ([emoji key](https://allcontributors.org/docs/e
 
 <!-- markdownlint-enable -->
 <!-- prettier-ignore-end -->
+
```
Comment (anteprimorac) [inline README.md:273]: @merelinguist this change was introduced by husky pre-commit hook

**Comment by anteprimorac** [comment]
@flybayer if you agree, we can merge this as it is and gather more info on what to provide regarding validation, sanitization and casting of params and queries.

**Comment by flybayer** [comment]
@anteprimorac let's wait until we resolve our discussion in #530. I just made another comment.

And if we add `useParamNumber()` and friends, I think it'd be better to use `useParams()` instead of `useRouterParams()` so all the param hooks are the same, and it's only query that is different. Additionally, all (or most) of other routers have `useParams()` so this will be more natural/intuitive for many folks.

**Comment by flybayer** [comment]
@anteprimorac what do you think about also changing `useRouter()` to have `params` and `query` to align with these new hooks?

**Current:**
```tsx
// Everything lumped in query
const {query} = useRouter()
```

**Idea:**
```tsx
// Separate route params from query string items
const {params, query} = useRouter()
```

This would be a breaking change for existing blitz apps

**Comment by anteprimorac** [comment]
@flybayer that’s an excellent idea. We could provide `useRouter` as a wrapper around the hook from next.js

**Comment by flybayer** [comment]
Yes exactly. I think we should do that

**Pull request merged**